### PR TITLE
Handles more argument cases for FB global functions

### PR DESCRIPTION
**Pull request by trueadm**
Release notes: none

Applied a better way for building FB abstract global functions where we might take arguments that are more complex than strings.

Fixes https://github.com/facebook/prepack/issues/1401

### CODE REVIEW

File: src/intrinsics/fb-www/fb-mocks.js [hunk 1]
```diff
@@ -73,18 +68,12 @@ const fbMagicGlobalObjects = ["JSResource", "Bootloader"];
...
-        functionValue = createAbstract(realm, "function", value);
-        realm.fbLibraries.other.set(value, functionValue);
-      }
-      invariant(functionValue instanceof AbstractValue);
+    value: new NativeFunctionValue(realm, functionName, functionName, 0, (context, args) => {
+      let functionValue = createAbstract(realm, "function");
+      functionValue.args = args;
+      // build the magic global function up manually from the arguments we pass it
+      // i.e. cx(...arg)
+      functionValue._buildNode = _args => t.callExpression(t.identifier(functionName), ((_args: any): Array<any>));
```
Comment (gaearon) [inline src/intrinsics/fb-www/fb-mocks.js]: This looks like a private API. Would be nice to verify it won't break with someone. `: any` probably isn't great either.
Maybe there's a supported way to do this? @NTillmann

**Comment by gaearon** [comment]
We'll probably want some tests too so we don't regress on this.

**Comment by trueadm** [comment]
I added tests to this. Well modified existing as per feedback. It’s hacks but then so are these non standards magic globals. Otherwise I’d use the standard abstract API, but it would not keep the standard we need internally.

File: src/intrinsics/fb-www/fb-mocks.js [hunk 2]
```diff
@@ -73,18 +68,12 @@ const fbMagicGlobalObjects = ["JSResource", "Bootloader"];
...
-
-      if (realm.fbLibraries.other.has(value)) {
-        functionValue = realm.fbLibraries.other.get(value);
-      } else {
-        functionValue = createAbstract(realm, "function", value);
-        realm.fbLibraries.other.set(value, functionValue);
-      }
-      invariant(functionValue instanceof AbstractValue);
+    value: new NativeFunctionValue(realm, functionName, functionName, 0, (context, args) => {
+      let functionValue = createAbstract(realm, "function");
```
Comment (gaearon) [inline src/intrinsics/fb-www/fb-mocks.js]: My impression is that AbstractValue constructor gives you more control. Maybe if you use that you won’t have to assign a private property.

Comment (trueadm) [reply src/intrinsics/fb-www/fb-mocks.js]: I tried that originally but I couldn’t get it to serialize right

**Comment by gaearon** [comment]
>I added tests to this. Well modified existing as per feedback

Hmm I don’t see, did you push?

**Comment by trueadm** [comment]
@gaearon Sorry, I added the updated test.

File: src/intrinsics/fb-www/fb-mocks.js (hunk 1 above)
Comment (trueadm) [reply src/intrinsics/fb-www/fb-mocks.js]: There is a way via creating the `AbstractValue` via the constructor route, but that means more boilerplate code than this route. This is already a somewhat hacky route as what we're doing isn't really "normal" – in that we want to serialize something in a very strict manner.

**Comment by gaearon** [comment]
Is there any way we can tweak the test that would ensure the output is actually the same as we expect? For example `cx()` didn't throw before with multiple arguments, it just silently emitted one argument. It would be nice to verify it works as intended.

**Comment by trueadm** [comment]
@gaearon We could change the code somewhat to throw, but I manually checked the output in this case.

**Comment by gaearon** [comment]
I think that if we use a private API then we need a test that will fail when that API no longer works. Just verifying the output once is not enough. In general, this seems to be a crucial part of www-specific mocks so I do think a test that checks the output is necessary.

File: src/intrinsics/fb-www/fb-mocks.js [hunk 3]
```diff
@@ -69,12 +75,16 @@ const fbMagicGlobalObjects = ["JSResource", "Bootloader"];
...
-      return functionValue;
+      let types = new TypesDomain(FunctionValue);
+      let values = new ValuesDomain();
+      invariant(context.$Realm.generator);
+      return context.$Realm.generator.derive(
+        types,
+        values,
+        args,
+        _args => t.callExpression(t.identifier(functionName), ((_args: any): Array<any>)),
+        { skipInvariant: true }
```
Comment (gaearon) [inline src/intrinsics/fb-www/fb-mocks.js]: What invariant does this skip and why?

Comment (trueadm) [reply src/intrinsics/fb-www/fb-mocks.js]: That was for debugging. Removed :)

**Pull request closed without merging**